nvc/nvc.go

```diff
@@ -0,0 +1,147 @@
+// Package nvc implements a parser for nvc archive files.
+//
+// The format of an archive is described below. All header values are stored on disk as little endian.
+//
+//   1. NVC magic bytes (8 bytes)
+//   2. Number of entries in the archive's table of contents (32-bit little endian unsigned integer)
+//   3. Table of Contents entries (192 * n bytes, where n is the previously-specified number of entries)
+package nvc
+
+import (
+	"compress/zlib"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+)
+
+// EntryFlags describes how a file is stored on disk
+type EntryFlags uint32
+
+const (
+	Uncompressed EntryFlags = 0
+	Compressed   EntryFlags = 1
+	Encrypted    EntryFlags = 3
+
+	magic = "nvc1d\x00\x00\x00"
+)
+
+var ErrNoMagicFound error = errors.New("nvc magic bytes not found")
+
+// TocEntry is a file entry in the table of contents.
+type TocEntry struct {
+	Hash      Hash       // 64-bit FNV-1a hash of the file's path on disk
+	Offset    uint32     // File's offset (in bytes) from the start of the nvc file
+	RawLength uint32     // Length (in bytes) of the file after it has been extracted
+	Length    uint32     // Length (in bytes) of file as it is stored in the nvc file
+	Flags     EntryFlags // Indicates whether file is compressed or encrypted (TODO: determine if this is a bitmask)
+}
+
+// ReadToc parses r as an NVC archive and returns its table of contents.
+func ReadToc(r io.Reader) ([]TocEntry, error) {
+	if magicErr := readMagic(r); magicErr != nil {
+		return nil, magicErr
+	}
+	count, countErr := readCount(r)
+	if countErr != nil {
+		return nil, countErr
+	}
+
+	entries := make([]TocEntry, count)
+
+	for i := range entries {
+		entry, eErr := readEntry(r)
+		if eErr != nil {
+			return nil, eErr
+		}
+		entries[i] = entry
+	}
+
+	return entries, nil
+
+}
+
+// Data returns from r the file that t is an entry for.
+func (t TocEntry) Data(r io.ReadSeeker) ([]byte, error) {
+	_, err := r.Seek(int64(t.Offset), 0)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	var reader io.Reader = r
+
+	switch t.Flags {
+	case Uncompressed:
+		// Do nothing
+	case Compressed:
+		reader, err = zlib.NewReader(r)
+		if err != nil {
+			return []byte{}, err
+		}
+	default:
+		return []byte{}, errors.New("unsupported")
+	}
+
+	data := make([]byte, t.RawLength)
+	_, err = io.ReadFull(reader, data)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return data, nil
+}
```

Comment:
throughout this func it i see returing an empty byte slice on error (`[]byte{}`), should this return `nil` instead?

Reply:
It's not a huge difference, but, yeah, `nil` _is_ preferable. Fixed in b8524d86d5cc79b382e7058dfd4f93b547f645bb